File: frontend/src/lib/catalog.ts
// List endpoints for catalog + stock screens, built on the shared request layer.
import { api } from "@/lib/api";
import type {
  Criticality,
  InventoryRow,
  Page,
  Product,
  ProductStatus,
  Supplier,
  SupplierStatus,
  Warehouse,
} from "@/types/api";

function qs(params: object): string {
  const sp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === "" || v === false) continue;
    sp.set(k, String(v));
  }
  const s = sp.toString();
  return s ? `?${s}` : "";
}

export interface ProductListParams {
  search?: string;
  status?: string;
  supplier_id?: string;
  page?: number;
  page_size?: number;
}

export interface SupplierListParams {
  search?: string;
  status?: string;
  page?: number;
  page_size?: number;
}

export interface WarehouseListParams {
  active_only?: boolean;
  page?: number;
  page_size?: number;
}

export interface InventoryListParams {
  warehouse_id?: string;
  product_id?: string;
  page?: number;
  page_size?: number;
}

export interface SupplierInput {
  name: string;
  contact_person?: string | null;
  email?: string | null;
  phone?: string | null;
  country?: string | null;
  currency: string;
  payment_terms?: string | null;
  default_lead_time_days: number;
  status: SupplierStatus;
}

export interface WarehouseInput {
  code: string;
  name: string;
  address?: string | null;
  branch_id?: string | null;
  is_active: boolean;
}

export interface ProductInput {
  sku: string;
  name: string;
  barcode?: string | null;
  description?: string | null;
  primary_supplier_id?: string | null;
  cost_price: string;
  selling_price: string;
  units_per_carton: number;
  moq: number;
  lead_time_days: number;
  reorder_point?: number | null;
  safety_stock?: number | null;
  status: ProductStatus;
  // Category/Brand by name (get-or-created server-side, matching the import flow).
  category?: string | null;
  brand?: string | null;
  // Product intelligence profile
  unit_of_measure?: string | null;
  currency?: string | null;
  commodity_tags?: string[];
  country_of_origin?: string | null;
  criticality?: Criticality;
  strategic_item?: boolean;
  alternate_supplier_available?: boolean;
}

export const catalogApi = {
  // ---- reads ----
  products: (p: ProductListParams = {}) => api.get<Page<Product>>(`/products${qs(p)}`),
  suppliers: (p: SupplierListParams = {}) => api.get<Page<Supplier>>(`/suppliers${qs(p)}`),
  warehouses: (p: WarehouseListParams = {}) => api.get<Page<Warehouse>>(`/warehouses${qs(p)}`),
  inventory: (p: InventoryListParams = {}) => api.get<Page<InventoryRow>>(`/inventory${qs(p)}`),

  // ---- suppliers ----
  createSupplier: (body: SupplierInput) => api.post<Supplier>("/suppliers", body),
  updateSupplier: (id: string, body: Partial<SupplierInput>) =>
    api.patch<Supplier>(`/suppliers/${id}`, body),
  deleteSupplier: (id: string) => api.del<void>(`/suppliers/${id}`),

  // ---- warehouses ----
  createWarehouse: (body: WarehouseInput) => api.post<Warehouse>("/warehouses", body),
  updateWarehouse: (id: string, body: Partial<WarehouseInput>) =>
    api.patch<Warehouse>(`/warehouses/${id}`, body),
  deleteWarehouse: (id: string) => api.del<void>(`/warehouses/${id}`),

  // ---- products ----
  createProduct: (body: ProductInput) => api.post<Product>("/products", body),
  updateProduct: (id: string, body: Partial<ProductInput>) =>
    api.patch<Product>(`/products/${id}`, body),
  deleteProduct: (id: string) => api.del<void>(`/products/${id}`),
};

File: backend/app/models/catalog.py
"""Catalog models: categories, brands, suppliers, products, supplier_products."""
from __future__ import annotations

import datetime as dt
import uuid
from decimal import Decimal

from sqlalchemy import (
    CHAR,
    Boolean,
    ForeignKey,
    Integer,
    Numeric,
    Text,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB, TIMESTAMP
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base

_UUID = PGUUID(as_uuid=True)


class Category(Base):
    __tablename__ = "categories"

    id: Mapped[uuid.UUID] = mapped_column(_UUID, primary_key=True, server_default=text("gen_random_uuid()"))
    tenant_id: Mapped[uuid.UUID] = mapped_column(_UUID, ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(Text, nullable=False)
    parent_id: Mapped[uuid.UUID | None] = mapped_column(_UUID, ForeignKey("categories.id", ondelete="SET NULL"), nullable=True)
    created_at: Mapped[dt.datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=text("now()"))


class Brand(Base):
    __tablename__ = "brands"

    id: Mapped[uuid.UUID] = mapped_column(_UUID, primary_key=True, server_default=text("gen_random_uuid()"))
    tenant_id: Mapped[uuid.UUID] = mapped_column(_UUID, ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(Text, nullable=False)
    created_at: Mapped[dt.datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=text("now()"))


class Supplier(Base):
    __tablename__ = "suppliers"

    id: Mapped[uuid.UUID] = mapped_column(_UUID, primary_key=True, server_default=text("gen_random_uuid()"))
    tenant_id: Mapped[uuid.UUID] = mapped_column(_UUID, ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(Text, nullable=False)
    code: Mapped[str | None] = mapped_column(Text, nullable=True)
    contact_person: Mapped[str | None] = mapped_column(Text, nullable=True)
    email: Mapped[str | None] = mapped_column(Text, nullable=True)
    phone: Mapped[str | None] = mapped_column(Text, nullable=True)
    address: Mapped[str | None] = mapped_column(Text, nullable=True)
    country: Mapped[str | None] = mapped_column(Text, nullable=True)
    currency: Mapped[str] = mapped_column(CHAR(3), nullable=False, server_default=text("'USD'"))
    payment_terms: Mapped[str | None] = mapped_column(Text, nullable=True)
    default_lead_time_days: Mapped[int] = mapped_column(Integer, nullable=False, server_default=text("30"))
    status: Mapped[str] = mapped_column(Text, nullable=False, server_default=text("'active'"))
    deleted_at: Mapped[dt.datetime | None] = mapped_column(TIMESTAMP(timezone=True), nullable=True)
    created_at: Mapped[dt.datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=text("now()"))
    updated_at: Mapped[dt.datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=text("now()"))


class Product(Base):
    __tablename__ = "products"

    id: Mapped[uuid.UUID] = mapped_column(_UUID, primary_key=True, server_default=text("gen_random_uuid()"))
    tenant_id: Mapped[uuid.UUID] = mapped_column(_UUID, ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    sku: Mapped[str] = mapped_column(Text, nullable=False)
    barcode: Mapped[str | None] = mapped_column(Text, nullable=True)
    name: Mapped[str] = mapped_column(Text, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    category_id: Mapped[uuid.UUID | None] = mapped_column(_UUID, ForeignKey("categories.id", ondelete="SET NULL"), nullable=True)
    brand_id: Mapped[uuid.UUID | None] = mapped_column(_UUID, ForeignKey("brands.id", ondelete="SET NULL"), nullable=True)
    primary_supplier_id: Mapped[uuid.UUID | None] = mapped_column(_UUID, ForeignKey("suppliers.id", ondelete="SET NULL"), nullable=True)
    cost_price: Mapped[Decimal] = mapped_column(Numeric(18, 4), nullable=False, server_default=text("0"))
    selling_price: Mapped[Decimal] = mapped_column(Numeric(18, 4), nullable=False, server_default=text("0"))
    units_per_carton: Mapped[int] = mapped_column(Integer, nullable=False, server_default=text("1"))
    moq: Mapped[int] = mapped_column(Integer, nullable=False, server_default=text("0"))
    lead_time_days: Mapped[int] = mapped_column(Integer, nullable=False, server_default=text("30"))
    weight_per_unit: Mapped[Decimal | None] = mapped_column(Numeric(18, 4), nullable=True)
    volume_per_unit: Mapped[Decimal | None] = mapped_column(Numeric(18, 6), nullable=True)
    weight_per_carton: Mapped[Decimal | None] = mapped_column(Numeric(18, 4), nullable=True)
    volume_per_carton: Mapped[Decimal | None] = mapped_column(Numeric(18, 6), nullable=True)
    cartons_per_pallet: Mapped[int | None] = mapped_column(Integer, nullable=True)
    reorder_point: Mapped[int | None] = mapped_column(Integer, nullable=True)
    safety_stock: Mapped[int | None] = mapped_column(Integer, nullable=True)
    # --- Product Intelligence Profile (migration 0009) ---
    # Consumed by the forecast, risk, procurement, intelligence, and AI engines.
    commodity_tags: Mapped[list] = mapped_column(JSONB, nullable=False, server_default=text("'[]'::jsonb"))
    country_of_origin: Mapped[str | None] = mapped_column(Text, nullable=True)
    transport_mode: Mapped[str | None] = mapped_column(Text, nullable=True)
    criticality: Mapped[str] = mapped_column(Text, nullable=False, server_default=text("'medium'"))
    supplier_dependency: Mapped[str | None] = mapped_column(Text, nullable=True)
    demand_type: Mapped[str | None] = mapped_column(Text, nullable=True)
    substitutability: Mapped[str | None] = mapped_column(Text, nullable=True)
    # --- Strategic flags (migration 0013) — consumed by the risk/forecast/AI engines ---
    strategic_item: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default=text("false"))
    alternate_supplier_available: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default=text("false"))
    # --- Unit of measure + product-level currency (import framework, migration 0011) ---
    # currency NULL => fall back to the tenant base_currency.
    unit_of_measure: Mapped[str | None] = mapped_column(Text, nullable=True)
    currency: Mapped[str | None] = mapped_column(CHAR(3), nullable=True)
    # Set when a row created this product via an import; powers rollback. NULL otherwise.
    created_by_import_job_id: Mapped[uuid.UUID | None] = mapped_column(_UUID, nullable=True)
    status: Mapped[str] = mapped_column(Text, nullable=False, server_default=text("'active'"))
    deleted_at: Mapped[dt.datetime | None] = mapped_column(TIMESTAMP(timezone=True), nullable=True)
    created_at: Mapped[dt.datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=text("now()"))
    updated_at: Mapped[dt.datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=text("now()"))


class SupplierProduct(Base):
    __tablename__ = "supplier_products"

    id: Mapped[uuid.UUID] = mapped_column(_UUID, primary_key=True, server_default=text("gen_random_uuid()"))
    tenant_id: Mapped[uuid.UUID] = mapped_column(_UUID, ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    supplier_id: Mapped[uuid.UUID] = mapped_column(_UUID, ForeignKey("suppliers.id", ondelete="CASCADE"), nullable=False)
    product_id: Mapped[uuid.UUID] = mapped_column(_UUID, ForeignKey("products.id", ondelete="CASCADE"), nullable=False)
    supplier_sku: Mapped[str | None] = mapped_column(Text, nullable=True)
    cost_price: Mapped[Decimal] = mapped_column(Numeric(18, 4), nullable=False, server_default=text("0"))
    currency: Mapped[str] = mapped_column(CHAR(3), nullable=False, server_default=text("'USD'"))
    moq: Mapped[int] = mapped_column(Integer, nullable=False, server_default=text("0"))
    lead_time_days: Mapped[int] = mapped_column(Integer, nullable=False, server_default=text("30"))
    units_per_carton: Mapped[int | None] = mapped_column(Integer, nullable=True)
    is_preferred: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default=text("false"))
    created_at: Mapped[dt.datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=text("now()"))
    updated_at: Mapped[dt.datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=text("now()"))
